src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs

```diff
@@ -12,216 +13,194 @@ namespace System
+                return string.Create(end - start, (str, start), static (buffer, state) =>
+                {
+                    int newLength = state.str.AsSpan(state.start, buffer.Length).ToLowerInvariant(buffer);
+                    Debug.Assert(newLength == buffer.Length);
+                });
+            }
+
+            string res = str.Substring(start, end - start);
 
-            if (res == Localhost || res == Loopback)
+            if (res is Localhost or Loopback)
             {
                 loopback = true;
                 return Localhost;
             }
+
             return res;
         }
 
-        //
-        // IsValid
-        //
-        //  Determines whether a string is a valid domain name
-        //
-        //      subdomain -> <label> | <label> "." <subdomain>
-        //
-        // Inputs:
-        //  - name as Name to test
-        //  - starting position
-        //  - ending position
-        //
-        // Outputs:
-        //  The end position of a valid domain name string, the canonical flag if found so
-        //
-        // Returns:
-        //  bool
-        //
-        //  Remarks: Optimized for speed as a most common case,
-        //           MUST NOT be used unless all input indexes are verified and trusted.
-        //
-
-        internal static unsafe bool IsValid(char* name, int pos, ref int returnedEnd, ref bool notCanonical, bool notImplicitFile)
+        public static bool IsValid(ReadOnlySpan<char> hostname, bool iri, bool notImplicitFile, out int length)
         {
-            char* curPos = name + pos;
-            char* newPos = curPos;
-            char* end = name + returnedEnd;
-            for (; newPos < end; ++newPos)
+            int invalidCharOrDelimiterIndex = iri
+                ? hostname.IndexOfAny(s_iriInvalidAsciiChars)
+                : hostname.IndexOfAnyExcept(s_validChars);
+
+            if (invalidCharOrDelimiterIndex >= 0)
             {
-                char ch = *newPos;
-                if (ch > 0x7f) return false;    // not ascii
-                if (ch < 'a') // Optimize for lower-case letters, which make up the majority of most Uris, and which are all greater than symbols checked for below
+                char c = hostname[invalidCharOrDelimiterIndex];
+
+                if (c == '/' || c == '\\' || (notImplicitFile && (c == ':' || c == '?' || c == '#')))
                 {
-                    if (ch == '/' || ch == '\\' || (notImplicitFile && (ch == ':' || ch == '?' || ch == '#')))
-                    {
-                        end = newPos;
-                        break;
-                    }
+                    hostname = hostname.Slice(0, invalidCharOrDelimiterIndex);
+                }
+                else
+                {
+                    length = 0;
+                    return false;
                 }
             }
 
-            if (end == curPos)
+            length = hostname.Length;
```

Comment:
Either move this to after the IsEmpty check or change the IsEmpty check to be length == 0?